File: FootballExplorer/app/src/main/java/fr/a42consulting/footballexplorer/View/TeamsSearchActivity.java
package fr.a42consulting.footballexplorer.View;

import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.GridView;
import android.widget.SearchView;
import android.widget.Toast;

import java.util.ArrayList;

import fr.a42consulting.footballexplorer.Model.TeamsSearchTeam;
import fr.a42consulting.footballexplorer.Presenter.ITeamsSearchPresenter;
import fr.a42consulting.footballexplorer.Presenter.TeamsSearchPresenter;
import fr.a42consulting.footballexplorer.R;

public class TeamsSearchActivity extends AppCompatActivity implements ITeamsSearchView {

    SearchView searchField;
    GridView resultsGrid;

    ITeamsSearchPresenter teamsSearchPresenter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search_teams);

        //Init view
        searchField = findViewById(R.id.searchField);
        resultsGrid = findViewById(R.id.resultsGrid);

        //Init
        teamsSearchPresenter = new TeamsSearchPresenter(this);

        searchField.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                teamsSearchPresenter.onSearch(query);
                searchField.clearFocus();
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });

        resultsGrid.setOnItemClickListener(new GridView.OnItemClickListener() {

            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                teamsSearchPresenter.onSelectTeam(position);
            }
        });
    }

    @Override
    public void onSearchResult(ArrayList<TeamsSearchTeam> teams) {
        TeamSearchGridAdapter gridAdapter = new TeamSearchGridAdapter(this, R.layout.grid_team_element);
        resultsGrid.setAdapter(gridAdapter);
        gridAdapter.addAll(teams);
        gridAdapter.notifyDataSetChanged();
        if (teams.size() == 0) {
            onMessage("La recherche n’a donné aucun résultat.");
        } else {
            onMessage(teams.size() + " résultat(s) trouvé(s)");
        }
    }

    @Override
    public void onShowTeam(String team) {
        Intent intent = new Intent(TeamsSearchActivity.this, PlayersListActivity.class);
        Bundle b = new Bundle();
        b.putString(PlayersListActivity.TEAM_NAME, team);
        intent.putExtras(b);
        startActivity(intent);
    }

    @Override
    public void onMessage(String message) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
    }
}

File: FootballExplorer_kotlin/app/src/main/java/fr/a42consulting/footballexplorer/view/TeamsSearchActivity.kt
package fr.a42consulting.footballexplorer.view

import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.widget.SearchView
import android.widget.Toast
import com.google.android.flexbox.FlexboxLayoutManager
import com.google.android.flexbox.JustifyContent

import java.util.ArrayList

import fr.a42consulting.footballexplorer.model.TeamsSearchTeam
import fr.a42consulting.footballexplorer.presenter.ITeamsSearchPresenter
import fr.a42consulting.footballexplorer.presenter.TeamsSearchPresenter
import fr.a42consulting.footballexplorer.R
import fr.a42consulting.footballexplorer.view.adapter.TeamSearchGridAdapter
import kotlinx.android.synthetic.main.activity_search_teams.*

class TeamsSearchActivity : AppCompatActivity(), ITeamsSearchView {
    private lateinit var teamsSearchPresenter: ITeamsSearchPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_search_teams)

        val layoutManager = FlexboxLayoutManager(this)
        layoutManager.justifyContent = JustifyContent.CENTER

        resultsGrid.layoutManager = layoutManager

        if (lastCustomNonConfigurationInstance == null) {
            teamsSearchPresenter = TeamsSearchPresenter(this)
        } else {
            teamsSearchPresenter = lastCustomNonConfigurationInstance as ITeamsSearchPresenter
            teamsSearchPresenter.resetView(this)
        }

        searchField.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                teamsSearchPresenter.onSearch(query)
                searchField.clearFocus()
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                return false
            }
        })
    }

    override fun onSearchResult(teams: ArrayList<TeamsSearchTeam>, isNewSearch: Boolean) {
        val gridAdapter = TeamSearchGridAdapter(this, teams) {
            teamsSearchPresenter.onSelectTeam(it)
        }
        resultsGrid.adapter = gridAdapter
        gridAdapter.notifyDataSetChanged()
        if (isNewSearch) {
            if (teams.size == 0) {
                onMessage(getString(R.string.teams_found_no))
            } else {
                onMessage(teams.size.toString() + getString(R.string.teams_found))
            }
        }
    }

    override fun onSearchFailed() {
        onMessage(getString(R.string.teams_found_failed))
    }

    override fun onShowTeam(team: String) {
        val intent = Intent(this@TeamsSearchActivity, PlayersListActivity::class.java)
        val b = Bundle()
        b.putString(PlayersListActivity.TEAM_NAME, team)
        intent.putExtras(b)
        startActivity(intent)
    }

    override fun onMessage(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
    }

    override fun onRetainCustomNonConfigurationInstance(): ITeamsSearchPresenter {
        return teamsSearchPresenter
    }

    override fun onDestroy() {
        super.onDestroy()
        teamsSearchPresenter.resetView(null)
        searchField.setOnQueryTextListener(null)
        resultsGrid.adapter = null
    }
}
